src/main/java/dev/openfeature/sdk/providers/memory/InMemoryProvider.java

```diff
@@ -0,0 +1,184 @@
+    }
+
+    @Override
+    public ProviderEvaluation<String> getStringEvaluation(String key, String defaultValue,
+            EvaluationContext evaluationContext) {
+        return getEvaluation(key, defaultValue, evaluationContext, String.class);
+    }
+
+    @Override
+    public ProviderEvaluation<Integer> getIntegerEvaluation(String key, Integer defaultValue,
+            EvaluationContext evaluationContext) {
+        return getEvaluation(key, defaultValue, evaluationContext, Integer.class);
+    }
+
+    @Override
+    public ProviderEvaluation<Double> getDoubleEvaluation(String key, Double defaultValue,
+            EvaluationContext evaluationContext) {
+        return getEvaluation(key, defaultValue, evaluationContext, Double.class);
+    }
+
+    @SneakyThrows
+    @Override
+    public ProviderEvaluation<Value> getObjectEvaluation(String key, Value defaultValue,
+            EvaluationContext evaluationContext) {
+        return getEvaluation(key, defaultValue, evaluationContext, Value.class);
+    }
+
+    private <T> ProviderEvaluation<T> getEvaluation(
+            String key, T defaultValue, EvaluationContext evaluationContext, Class<?> expectedType
+    ) throws OpenFeatureError {
+        if (!ProviderState.READY.equals(state)) {
+            ErrorCode errorCode = ErrorCode.PROVIDER_NOT_READY;
+            if (ProviderState.ERROR.equals(state)) {
+                errorCode = ErrorCode.GENERAL;
+            }
+            return ProviderEvaluation.<T>builder()
+                .errorCode(errorCode)
+                .reason(errorCode.name())
+                .value(defaultValue)
+                .build();
+        }
+        Flag<?> flag = flags.get(key);
+        if (flag == null) {
+            return ProviderEvaluation.<T>builder()
+                .value(defaultValue)
+                .reason(Reason.ERROR.toString())
+                .errorMessage(ErrorCode.FLAG_NOT_FOUND.name())
+                .errorCode(ErrorCode.FLAG_NOT_FOUND)
+                .build();
+        }
+        T value;
+        if (flag.getContextEvaluator() != null) {
+            value = (T) flag.getContextEvaluator().evaluate(flag, evaluationContext);
+        } else if (!expectedType.isInstance(flag.getVariants().get(flag.getDefaultVariant()))) {
+            return ProviderEvaluation.<T>builder()
+                .value(defaultValue)
+                .variant(flag.getDefaultVariant())
+                .reason(Reason.ERROR.toString())
+                .errorMessage(ErrorCode.TYPE_MISMATCH.name())
+                .errorCode(ErrorCode.TYPE_MISMATCH)
+                .build();
+        } else {
+            value = (T) flag.getVariants().get(flag.getDefaultVariant());
+        }
+        return ProviderEvaluation.<T>builder()
+            .value(value)
+            .variant(flag.getDefaultVariant())
+            .reason(Reason.STATIC.toString())
+            .build();
+    }
+
+    private static Value objectToValue(Object object) {
```

Comment:
I think this should be exposed in `Value` class itself. This will help other provider implementations